controllers/openstackcluster_controller.go

```diff
@@ -206,22 +190,30 @@ func deleteBastion(log logr.Logger, osProviderClient *gophercloud.ProviderClient
 		return err
 	}
 
-	if openStackCluster.Status.Bastion != nil {
-		if err = computeService.DeleteBastion(openStackCluster, openStackCluster.Status.Bastion.ID); err != nil {
-			return errors.Errorf("failed to delete bastion: %v", err)
-		}
-		if err = networkingService.DeleteFloatingIP(openStackCluster, openStackCluster.Status.Bastion.FloatingIP); err != nil {
-			return errors.Errorf("failed to delete floating IP: %v", err)
-		}
+	clusterName := fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)
+	bastionName := getBastionName(clusterName)
+
+	instance, err := computeService.InstanceExists(bastionName)
+	if err != nil {
+		return err
+	}
+	if instance == nil {
 		openStackCluster.Status.Bastion = nil
+		return nil
 	}
 
-	if openStackCluster.Status.BastionSecurityGroup != nil {
-		if err = networkingService.DeleteSecurityGroups(openStackCluster, openStackCluster.Status.BastionSecurityGroup); err != nil {
-			return errors.Errorf("failed to delete security group: %v", err)
-		}
-		openStackCluster.Status.BastionSecurityGroup = nil
+	if err = computeService.DeleteBastion(openStackCluster, clusterName); err != nil {
+		return errors.Errorf("failed to delete bastion: %v", err)
+	}
+	if err = networkingService.DeleteFloatingIP(openStackCluster, instance.FloatingIP); err != nil {
```

Comment:
we will have problems with reentrancy with the current impl, e.g.:
* DeleteBastion works
* DeleteFloatingIP fails
* retry will return in l.201
* security group will never get deleted

=> I think we can only skip the bastion / floating ip deletion when we are 100% sure they are gone. Only the ensure the server is gone is not good enough.

Reply:
Thanks, I tried the rule, but totally wrong. I will fully update.